### PR TITLE
forward headers that are treated special into foxx as well.

### PR DESCRIPTION
**Pull request by dothebart**
### Scope & Purpose
Up to now, the http headers `accept`, `accept-encoding` and 

- [x] :hankey: Bugfix (requires CHANGELOG entry)
- [ ] :pizza: New feature (requires CHANGELOG entry, feature documentation and release notes)
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification
- [ ] :book: CHANGELOG entry made

#### Backports:

- [ ] No backports required
- [x] Backports required for: 3.7

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [x] GitHub issue: https://github.com/arangodb/arangodb/issues/12734
- [ ] Design document: 

### Testing & Verification

*(Please pick either of the following options)*

- [ ] This change is a trivial rework / code cleanup without any test coverage.
- [ ] The behavior in this PR was *manually tested*
- [ ] This change is already covered by existing tests, such as *(please describe tests)*.
- [x] This PR adds tests that were used to verify all changes:
  - [ ] Added new C++ **Unit tests**
  - [x] Added new **integration tests** (e.g. in shell_client)
  - [ ] Added new **resilience tests** (only if the feature is impacted by failovers)
- [ ] There are tests in an external testing repository:
- [ ] I ensured this code runs with ASan / TSan or other static verification tools

Link to Jenkins PR run:

### Documentation

> All new features should be accompanied by corresponding documentation. 
> Bugs and features should furthermore be documented in the CHANGELOG so that
> support, end users, and other developers have a concise overview. 

- [ ] Added entry to *Release Notes* 
- [ ] Added a new section in the *Manual* 
- [ ] Added a new section in the *HTTP API* 
- [ ] Added *Swagger examples* for the HTTP API  
- [ ] Updated license information in *LICENSES-OTHER-COMPONENTS.md* for 3rd party libraries

### CODE REVIEW

**Comment by dothebart** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/12830/

File: tests/js/client/shell/shell-foxx-api-spec.js [hunk 1]
```diff
@@ -87,22 +87,59 @@ describe('FoxxApi commit', function () {
...
     `).getExtra().stats.writesExecuted).to.equal(1);
 
+    let body;
     let result = arango.GET('/test/header-echo');
     expect(result.code).to.equal(404);
 
     result = arango.POST('/_api/foxx/commit', '');
     expect(result.code).to.equal(204);
+    [
+      // explicitely say we want utf8-json, since the server will add it:
```
Comment (jsteemann) [inline tests/js/client/shell/shell-foxx-api-spec.js:97]:
```suggestion
      // explicitly say we want utf8-json, since the server will add it:
```

File: tests/js/client/shell/shell-foxx-api-spec.js [hunk 2]
```diff
@@ -87,22 +87,59 @@ describe('FoxxApi commit', function () {
...
+          }
+        }
+        expect(body[key]).to.equal(headers[key]);
+        
+      });
+
+    });
 
-    result = arango.GET_RAW('/test/header-echo', { origin: origin });
+    // sending content-type json actually requries to post something:
```
Comment (jsteemann) [inline tests/js/client/shell/shell-foxx-api-spec.js:128]:
```suggestion
    // sending content-type json actually requires to post something:
```

File: arangod/V8Server/v8-actions.cpp [hunk 3]
```diff
@@ -436,6 +436,62 @@ v8::Handle<v8::Object> TRI_RequestCppToV8(v8::Isolate* isolate,
...
+      break;
+    case ContentType::TEXT:    // text/plain
+      headers.insert(std::pair<std::string, std::string>(StaticStrings::ContentTypeHeader, StaticStrings::MimeTypeText));
+      break;
+    case ContentType::HTML:    // text/html
+      headers.insert(std::pair<std::string, std::string>(StaticStrings::ContentTypeHeader, StaticStrings::MimeTypeHtml));
+      break;
+    case ContentType::DUMP:    // application/x-arango-dump
+      headers.insert(std::pair<std::string, std::string>(StaticStrings::ContentTypeHeader, StaticStrings::MimeTypeDump));
+      break;
```
Comment (jsteemann) [inline arangod/V8Server/v8-actions.cpp:492]: Instead of `map::insert(std::pair<std::string, std::string>(key, value))`, can't we simply use `map::emplace(key, value`?

File: arangod/V8Server/v8-actions.cpp [hunk 4]
```diff
@@ -436,6 +436,62 @@ v8::Handle<v8::Object> TRI_RequestCppToV8(v8::Isolate* isolate,
...
+    break;
+  }
+  std::string const& acceptPlain = request->contentTypeResponsePlain();
+
+  if (!acceptPlain.empty()) {
+    headers.insert(std::pair<std::string, std::string>(StaticStrings::Accept, acceptPlain));
+  } else {
+    switch(request->contentTypeResponse()) {
+    case ContentType::UNSET:
+    case ContentType::CUSTOM:  // use Content-Type from _headers
```
Comment (jsteemann) [inline arangod/V8Server/v8-actions.cpp:454]: Mmh, this `switch` statement here has a different indentation than the one added a few lines below for the content type.

File: arangod/V8Server/v8-actions.cpp [hunk 5]
```diff
@@ -436,6 +436,62 @@ v8::Handle<v8::Object> TRI_RequestCppToV8(v8::Isolate* isolate,
...
+      break;
+    case ContentType::JSON:    // application/json
+      headers.insert(std::pair<std::string, std::string>(StaticStrings::Accept, StaticStrings::MimeTypeJson));
```
Comment (jsteemann) [inline arangod/V8Server/v8-actions.cpp:457]: let's use map::emplace here, too.

File: arangod/V8Server/v8-actions.cpp [hunk 6]
```diff
@@ -436,6 +436,62 @@ v8::Handle<v8::Object> TRI_RequestCppToV8(v8::Isolate* isolate,
...
+    case ContentType::HTML:    // text/html
+      headers.insert(std::pair<std::string, std::string>(StaticStrings::Accept, StaticStrings::MimeTypeHtml));
+      break;
+    case ContentType::DUMP:    // application/x-arango-dump
+      headers.insert(std::pair<std::string, std::string>(StaticStrings::Accept, StaticStrings::MimeTypeDump));
+      break;
+    }
+  }
+
+  switch(request->contentType()) {
```
Comment (jsteemann) [inline arangod/V8Server/v8-actions.cpp:474]:
```suggestion
  switch (request->contentType()) {
```

File: arangod/V8Server/v8-actions.cpp [hunk 7]
```diff
@@ -436,6 +436,62 @@ v8::Handle<v8::Object> TRI_RequestCppToV8(v8::Isolate* isolate,
...
+    headers.insert(std::pair<std::string, std::string>(StaticStrings::AcceptEncoding, StaticStrings::EncodingIdentity));
+    break;
+  case EncodingType::DEFLATE:
+    headers.insert(std::pair<std::string, std::string>(StaticStrings::AcceptEncoding, StaticStrings::EncodingDeflate));
+    break;
+  }
+  std::string const& acceptPlain = request->contentTypeResponsePlain();
+
+  if (!acceptPlain.empty()) {
+    headers.insert(std::pair<std::string, std::string>(StaticStrings::Accept, acceptPlain));
```
Comment (jsteemann) [inline arangod/V8Server/v8-actions.cpp:450]: Again, we can use empty here.

File: arangod/V8Server/v8-actions.cpp [hunk 8]
```diff
@@ -436,6 +436,62 @@ v8::Handle<v8::Object> TRI_RequestCppToV8(v8::Isolate* isolate,
   // intentional copy, as we will modify the headers later
   auto headers = request->headers();
 
+  switch (request->acceptEncoding()) {
+  case EncodingType::UNSET:
+    headers.insert(std::pair<std::string, std::string>(StaticStrings::AcceptEncoding, StaticStrings::EncodingIdentity));
```
Comment (jsteemann) [inline arangod/V8Server/v8-actions.cpp:441]: This change will mean that Foxx will now see a lot of requests coming in with `Accept-Encoding: identity`, however, this is not necessary what the client has actually sent.
Why is this change necessary?

**Comment by jsteemann** [comment]
In addition, I think we need to make a CHANGELOG entry, because with that change we will not be populating the Accept-Type header of incoming requests with `identity` in a lot of cases. This does not seem to be downwards-compatible, so we need to mention this in the CHANGELOG and add a paragraph to https://github.com/arangodb/docs/blob/main/3.8/release-notes-38.md

File: arangod/V8Server/v8-actions.cpp (hunk 4 above)
Comment (dothebart) [reply arangod/V8Server/v8-actions.cpp:454]: yes, its in an else-case. `'image/webp,text/html,application/x-html,*/*;q=0.8'`would become `text/html` here without the if and the storing of this exact header previously.

**Comment by dothebart** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/12835/

File: arangod/V8Server/v8-actions.cpp (hunk 8 above)
Comment (dothebart) [reply arangod/V8Server/v8-actions.cpp]: we could also store the original string, since there is more than `identity` and `deflate` in the world anyways. Then this can be the proper string.

**Review by mpoeter — APPROVED** [review]

**Comment by dothebart** [comment]
testrun blue. 
To be decided: what to do with `accept: identity|deflate` header? 
1) ommit on `identity` 
2) store actual string similar to the `accept-content-type` header, and use that
3) continue as is only add to CHANGELOG & documentation changelog

**Comment by pluma** [comment]
> Up to now, the http headers accept, accept-encoding and

The description of this issue seems to be missing a piece there.

**Comment by pluma** [comment]
> testrun blue.
> To be decided: what to do with `accept: identity|deflate` header?
> 
> 1. ommit on `identity`
> 2. store actual string similar to the `accept-content-type` header, and use that
> 3. continue as is only add to CHANGELOG & documentation changelog

My vote is for 2.

**Comment by dothebart** [comment]
hm, we have 4) - do nothing to `accept-encoding` as the algorithms leave it in that list, opposing the other headers. The tests now check for all variants of headers to become handled correctly, and no `accept-encoding` headers being added inbetween on random.

however, there seems to be another so far not yet identified issue: (as found in discussions with @pluma )
- foxx may use compressed files https://www.arangodb.com/docs/stable/foxx-reference-manifest.html 
- https://github.com/arangodb/arangodb/blob/devel/js/server/modules/%40arangodb/actions.js#L2033-L2047 - which depends on `accept-encoding: gzip` but won't work with `deflate` - though its basically the same?
- https://github.com/arangodb/arangodb/blob/devel/arangod/GeneralServer/RestHandler.cpp#L331 otoh will only react on `deflate` - not gzip.

currently there are no tests for neither behaviour, else they would have seen me breaking it. Letting `compressResponse()` also lose if we have `accept-encoding: gzip` may add missing features, but my cause foxx delivering `.gz` files end up with double-compressed content.

File: arangod/V8Server/v8-actions.cpp (hunk 8 above)
Comment (dothebart) [reply arangod/V8Server/v8-actions.cpp]: accept encoding wasn't filtered out as the other headers were - hence removed again.

File: 3rdParty/fuerte/src/message.cpp [hunk 9]
```diff
@@ -111,6 +111,10 @@ void RequestHeader::parseArangoPath(std::string const& p) {
 // class Message
 ///////////////////////////////////////////////
 
+      ContentEncoding Message::contentEncoding() const {
```
Comment (jsteemann) [inline 3rdParty/fuerte/src/message.cpp:114]: indentation?
```suggestion
ContentEncoding Message::contentEncoding() const {
```

File: 3rdParty/fuerte/src/types.cpp [hunk 10]
```diff
@@ -281,6 +281,38 @@ std::string to_string(ContentType type) {
...
+std::string to_string(ContentEncoding type) {
+  switch(type) {
+  default:
+  case ContentEncoding::Identity:
+    return fu_content_encoding_identity;
+  case ContentEncoding::Deflate:
+    return fu_content_encoding_deflate;
+  case ContentEncoding::Gzip:
+    return fu_content_encoding_gzip;
+  }
```
Comment (jsteemann) [inline 3rdParty/fuerte/src/types.cpp:313]:
```suggestion
  switch(type) {
    case ContentEncoding::Deflate:
      return fu_content_encoding_deflate;
    case ContentEncoding::Gzip:
      return fu_content_encoding_gzip;
    case ContentEncoding::Identity:
    default:
      return fu_content_encoding_identity;
  }
```

File: arangod/V8Server/v8-actions.cpp [hunk 11]
```diff
@@ -436,6 +436,54 @@ v8::Handle<v8::Object> TRI_RequestCppToV8(v8::Isolate* isolate,
...
+    case ContentType::TEXT:    // text/plain
+      headers.emplace(StaticStrings::Accept, StaticStrings::MimeTypeText);
+      break;
+    case ContentType::HTML:    // text/html
+      headers.emplace(StaticStrings::Accept, StaticStrings::MimeTypeHtml);
+      break;
+    case ContentType::DUMP:    // application/x-arango-dump
+      headers.emplace(StaticStrings::Accept, StaticStrings::MimeTypeDump);
+      break;
+    }
```
Comment (jsteemann) [inline arangod/V8Server/v8-actions.cpp:463]:
```suggestion
    switch (request->contentTypeResponse()) {
      case ContentType::UNSET:
      case ContentType::CUSTOM:  // use Content-Type from _headers
        break;
      case ContentType::JSON:    // application/json
        headers.emplace(StaticStrings::Accept, StaticStrings::MimeTypeJson);
        break;
      case ContentType::VPACK:   // application/x-velocypack
        headers.emplace(StaticStrings::Accept, StaticStrings::MimeTypeVPack);
        break;
      case ContentType::TEXT:    // text/plain
        headers.emplace(StaticStrings::Accept, StaticStrings::MimeTypeText);
        break;
      case ContentType::HTML:    // text/html
        headers.emplace(StaticStrings::Accept, StaticStrings::MimeTypeHtml);
        break;
      case ContentType::DUMP:    // application/x-arango-dump
        headers.emplace(StaticStrings::Accept, StaticStrings::MimeTypeDump);
        break;
    }
```

**Comment by dothebart** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/12905/

File: 3rdParty/fuerte/src/types.cpp [hunk 12]
```diff
@@ -281,6 +281,38 @@ std::string to_string(ContentType type) {
...
+                           fu_content_encoding_gzip) == 0) {
+      return ContentEncoding::Gzip;
+    } else if (val.compare(0, fu_content_encoding_deflate.size(),
+                           fu_content_encoding_deflate) == 0) {
+      return ContentEncoding::Deflate;
+    } else if (val.compare(0, fu_content_encoding_identity.size(),
+                           fu_content_encoding_identity) == 0) {
+      return ContentEncoding::Identity;
+    }
+    return ContentEncoding::Identity;
```
Comment (jsteemann) [inline 3rdParty/fuerte/src/types.cpp:301]:
```suggestion
ContentEncoding to_ContentEncoding(std::string const& val) {
  if (val.empty()) {
    return ContentEncoding::Identity;
  } else if (val.compare(0, fu_content_encoding_gzip.size(),
                         fu_content_encoding_gzip) == 0) {
    return ContentEncoding::Gzip;
  } else if (val.compare(0, fu_content_encoding_deflate.size(),
                         fu_content_encoding_deflate) == 0) {
    return ContentEncoding::Deflate;
  } else if (val.compare(0, fu_content_encoding_identity.size(),
                         fu_content_encoding_identity) == 0) {
    return ContentEncoding::Identity;
  }
  return ContentEncoding::Identity;
```

File: arangosh/Shell/V8ClientConnection.cpp [hunk 13]
```diff
@@ -1941,35 +1941,41 @@ v8::Local<v8::Value> V8ClientConnection::requestDataRaw(
...
+  auto setBinaryPayload = [isolate, context, responseBody, result]{
+        V8Buffer* buffer = V8Buffer::New
+          (isolate,
+           static_cast<char const*>(responseBody.data()),
+           responseBody.size());
+        auto bufObj = v8::Local<v8::Object>::New(isolate, buffer->_handle);
+        result->Set(context,
+                    TRI_V8_ASCII_STRING(isolate, "body"),
+                    bufObj).FromMaybe(false);
+                          };
```
Comment (jsteemann) [inline arangosh/Shell/V8ClientConnection.cpp:1954]: What happened to the indentation here?

File: arangosh/Shell/V8ClientConnection.cpp [hunk 14]
```diff
@@ -1941,35 +1941,41 @@ v8::Local<v8::Value> V8ClientConnection::requestDataRaw(
...
-      const char* bodyStr = reinterpret_cast<const char*>(responseBody.data());
-      v8::Local<v8::String> b = TRI_V8_PAIR_STRING(isolate, bodyStr, responseBody.size());
-      result->Set(context,
-                  TRI_V8_ASCII_STRING(isolate, "body"), b).FromMaybe(false);
+      if (response->contentEncoding() == fuerte::ContentEncoding::Identity) {
+        const char* bodyStr = reinterpret_cast<const char*>(responseBody.data());
+        v8::Local<v8::String> b = TRI_V8_PAIR_STRING(isolate, bodyStr, responseBody.size());
+        result->Set(context,
+                    TRI_V8_ASCII_STRING(isolate, "body"), b).FromMaybe(false);
+      } else setBinaryPayload();
```
Comment (jsteemann) [inline arangosh/Shell/V8ClientConnection.cpp:1972]:
```suggestion
      } else {
        setBinaryPayload();
      }
```

File: arangosh/Shell/V8ClientConnection.cpp [hunk 15]
```diff
@@ -1941,35 +1941,41 @@ v8::Local<v8::Value> V8ClientConnection::requestDataRaw(
...
-    V8Buffer* buffer = V8Buffer::New
-      (isolate,
-       static_cast<char const*>(responseBody.data()),
-       responseBody.size());
-    auto bufObj = v8::Local<v8::Object>::New(isolate, buffer->_handle);
-    result->Set(context,
-                TRI_V8_ASCII_STRING(isolate, "body"),
-                bufObj).FromMaybe(false);
-  }
+  } else setBinaryPayload();
```
Comment (jsteemann) [inline arangosh/Shell/V8ClientConnection.cpp:1978]:
```suggestion
  } else {
    setBinaryPayload();
  }
```

**Review by jsteemann — APPROVED** [review]
LGTM except indentation

File: arangosh/Shell/V8ClientConnection.cpp (hunk 13 above)
Comment (dothebart) [reply arangosh/Shell/V8ClientConnection.cpp]: fixeb.

**Comment by dothebart** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/12908/

File: arangosh/Shell/V8ClientConnection.cpp [hunk 16]
```diff
@@ -1941,35 +1941,42 @@ v8::Local<v8::Value> V8ClientConnection::requestDataRaw(
...
```
Comment (jsteemann) [inline arangosh/Shell/V8ClientConnection.cpp:1979]:
```suggestion
  } else {
    setBinaryPayload();
  }
```

File: arangosh/Shell/V8ClientConnection.cpp [hunk 17]
```diff
@@ -1941,35 +1941,42 @@ v8::Local<v8::Value> V8ClientConnection::requestDataRaw(
...
-      const char* bodyStr = reinterpret_cast<const char*>(responseBody.data());
-      v8::Local<v8::String> b = TRI_V8_PAIR_STRING(isolate, bodyStr, responseBody.size());
-      result->Set(context,
-                  TRI_V8_ASCII_STRING(isolate, "body"), b).FromMaybe(false);
+      if (response->contentEncoding() == fuerte::ContentEncoding::Identity) {
+        const char* bodyStr = reinterpret_cast<const char*>(responseBody.data());
+        v8::Local<v8::String> b = TRI_V8_PAIR_STRING(isolate, bodyStr, responseBody.size());
+        result->Set(context,
+                    TRI_V8_ASCII_STRING(isolate, "body"), b).FromMaybe(false);
+      } else setBinaryPayload();
```
Comment (jsteemann) [inline arangosh/Shell/V8ClientConnection.cpp:1973]:
```suggestion
      } else {
        setBinaryPayload();
      }
```

**Comment by dothebart** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr-windows/6476/ second windows test blue.
